### PR TITLE
feat: add support for lambda to create a MosaicJSON representaion of a STAC API Search result

### PR DESCRIPTION
**Pull request by philvarner**
Issue: https://github.com/MAAP-Project/ZenHub/issues/339

Deployed to https://qnz86q6bnb.execute-api.us-east-1.amazonaws.com/v1/mosaics

A request like this, containing a `stac_api_root`, a `username`, and the same parameters as accepted by the STAC API Item Search endpoint (`/search`):

```bash
curl -X "POST" "https://qnz86q6bnb.execute-api.us-east-1.amazonaws.com/v1/mosaics" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "stac_api_root": "https://earth-search.aws.element84.com/v0",
  "username": "philvarner",
  "datetime": "2021-04-20T00:00:00Z/2021-04-21T02:00:00Z",
  "collections": [
    "sentinel-s2-l2a-cogs"
  ],
  "bbox": [
    20,
    20,
    21,
    21
  ]
}'
```

will generate a mosaicjson endpoint and return a body with a Link to the tilejson for it:

```json
{
  "id": "philvarner.RNOyIRiuXOewxcwtbPuuaYWROmXtTxzy",
  "links": [
    {
      "href": "https://api.cogeo.xyz/mosaicjson/philvarner.RNOyIRiuXOewxcwtbPuuaYWROmXtTxzy/tilejson.json",
      "rel": "tilejson",
      "type": "application/json",
      "title": "TileJSON"
    }
  ]
}
```

This endpoint depends on cogeo-mosaic, which needs a newer version of rio-tiler (~=2.1.0) than dashboard api (2.0a.11). There are several incompatibilities between dashboard and rio-tiler >= 2.0.0, due to methods being removed in rio-tiler in 2.0.0 final that existed in the alphas. 🤷🏻 This shouldn't take but a few hours to fix, and it would be good in general to get rio-tiler upgraded.

This necessitated that the code be packaged separately and use a different setup.py, hence it all being a subdirectory. 

Ideally, we could update dashboard api to use a newer rio-tiler and just incorporate all of this code for the mosaic_api as another endpoint in dashboard api. Another option would be to move dashboard_api into sibling directory to mosaic_api instead of being in a parent of it.

Remaining Issues:
* more tests for external api failures
* timeouts weren't triggering sometimes when STAC API hung

### CODE REVIEW

File: mosaic_api/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Mosaic API
+
+An API to proxy STAC API searches into a MosaicJSON endpoint.
```
Comment (abarciauskas-bgse) [inline mosaic_api/README.md:3]: We might want to enumerate the endpoints we are creating (e.g. POST and GET mosaicjson) to make it clear we're not run a dynamic tiler for mosaicjson (yet!)

File: mosaic_api/README.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+}'
+```
+
+### Running the app with docker:
+
+```bash
+docker-compose up --build
+```
+
+## Contribution & Development
```
Comment (abarciauskas-bgse) [inline mosaic_api/README.md:73]: I think here and below can be removed? Or at least needs to be tailored for deploying the mosaic API, for example I don't think there is an install.sh script for this stack and it only deploys 2 stacks

File: mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+from asyncio import wait_for, create_task
+import asyncio
+
+from mosaic_api.core.config import MOSAIC_API_ROOT
+
+router = APIRouter()
+
+
+@router.get(
+    "/mosaics/{mosaic_id}",
```
Comment (abarciauskas-bgse) [inline mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py:22]: Naming convention changes requested: I think the API endpoint should be /mosaicjson and `mosaic_id` should be `layer` - this is consistent with https://api.cogeo.xyz/docs#/MosaicJSON

File: mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        task_token = create_task(retrieve_token(mosaic_request.username))
+
+        features = await wait_for(execute_stac_search(mosaic_request), 10)
+
+        mosaic_data = await wait_for(extract_mosaic_data(features), 20)
+
+        token = await wait_for(task_token, 5)
+
+        mosaic_id = await wait_for(create_mosaic_remote(layername, mosaic_request.username, token, mosaic_data), 5)
+
```
Comment (abarciauskas-bgse) [inline mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py:70]: Just a style thing but should we have all these blank lines?

**Review by abarciauskas-bgse — CHANGES_REQUESTED** [review]
This is looks great so far @philvarner - I'm also hoping we can get a review from @vincentsarago so tagging him to take a look since he is the mosaicjson spec author.

So does the CDK deployment code work?

**Comment by vincentsarago** [comment]
@philvarner 😭 we just updated a rio-tiler and cogeo-mosaic last week, I'm sorry for the trouble. I'll have a look at the PR as soon as I can!

File: mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py [hunk 5]
```diff
@@ -5,13 +5,13 @@
...
 from starlette import status
 from starlette.requests import Request
-import requests
 from cogeo_mosaic.mosaic import MosaicJSON
 from rio_tiler.io import COGReader
 from pystac_client import Client
 from typing import List, Optional
 from asyncio import wait_for, create_task
 import asyncio
+import aiohttp
```
Comment (vincentsarago) [inline mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py:14]: ❤️

File: mosaic_api/mosaic_api/__init__.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+"""mosaic_apiboard_api."""
+
+import pkg_resources
+
+version = pkg_resources.get_distribution(__package__).version
```
Comment (vincentsarago) [inline mosaic_api/mosaic_api/__init__.py:5]: we should avoid this, it makes startup really slow. 

We switched to hardcoded version in most of out lib (and use https://github.com/c4urself/bump2version to update versions accros the files)

File: mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+from starlette.requests import Request
+from cogeo_mosaic.mosaic import MosaicJSON
+from rio_tiler.io import COGReader
+from pystac_client import Client
+from typing import List, Optional
+from asyncio import wait_for, create_task
+import asyncio
+import aiohttp
+
+from mosaic_api.core.config import MOSAIC_API_ROOT
```
Comment (vincentsarago) [inline mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py:16]: I would suggest to use iSort to sort the module imports, but this is not mandatory

File: mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+)
+async def get_mosaic(
+        mosaic_id: str
+) -> Mosaic:
+    """Return mosaic info for the given mosaic ID"""
+    # todo: should this check if the mosaic actually exists, or just return the tilejson url?
+    if mosaic_id:
+        return _mosaic(mosaic_id)
+    else:
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="invalid mosaic ID")
```
Comment (vincentsarago) [inline mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py:36]: you can make path parameter required using `fastapi.Path(...)`  and also add description. As we do in https://github.com/developmentseed/titiler/blob/master/titiler/core/titiler/core/factory.py#L297

Comment (vincentsarago) [reply mosaic_api/mosaic_api/api/api_v1/endpoints/mosaics.py:36]: you can also mix path/query definition using `Dependencies` 

e.g: this is the dependency we use in api.cogeo.xyz.

```python
@dataclass
class MosaicParams:
    """Create mosaic path from args"""

    layer: str = Query(..., description="Mosaic Layer name ('{username}.{layer}')")

    def __post_init__(self,):
        """Define mosaic URL."""
        pattern = (
            r"^(?P<username>[a-zA-Z0-9-_]{1,32})\.(?P<layername>[a-zA-Z0-9-_]{1,32})$"
        )
        match = re.match(pattern, self.layer)
        if not match:
            raise HTTPException(
                status_code=400, detail=f"Invalid layer name: `{self.layer}`",
            )

        meta = match.groupdict()
        self.username = meta["username"]
        self.layername = meta["layername"]
        if mosaic_config.backend == "dynamodb://":
            self.url = f"{mosaic_config.backend}{mosaic_config.host}:{self.layer}"
        else:
            self.url = f"{mosaic_config.backend}{mosaic_config.host}/{self.username}/{self.layername}{mosaic_config.format}"
```

**Comment by abarciauskas-bgse** [comment]
I think we can close this in favor of using https://github.com/NASA-IMPACT/titiler/ for mosaic management and tiling right @philvarner ?

**Pull request closed without merging**